mqtt/mqtt-bridge/src/bridge.rs

```diff
@@ -1,482 +1,204 @@
-                .map(|stripped_topic| {
-                    if let Some(remote_prefix) = mapper.topic_settings.remote() {
-                        format!("{}/{}", remote_prefix, stripped_topic)
-                    } else {
-                        stripped_topic.to_string()
-                    }
-                })
-        })
+    fn format_key_value(topic: &TopicRule) -> (String, TopicRule) {
+        let key = if let Some(local) = topic.local() {
+            format!("{}/{}", local, topic.pattern().to_string())
+        } else {
+            topic.pattern().into()
+        };
+        (key, topic.clone())
     }
-}
 
-// TODO: implement for generic
-#[async_trait]
-impl EventHandler for MessageHandler<WakingMemoryStore> {
-    type Error = BridgeError;
+    pub async fn run(mut self) -> Result<(), BridgeError> {
+        info!("Starting {} bridge...", self.connection_settings.name());
 
-    async fn handle_event(&mut self, event: Event) -> Result<(), Self::Error> {
-        if let Event::Publication(publication) = event {
-            let ReceivedPublication {
-                topic_name,
-                qos,
-                retain,
-                payload,
-                dup: _,
-            } = publication;
-            let forward_publication = self.transform(topic_name.as_ref()).map(|f| Publication {
-                topic_name: f,
-                qos,
-                retain,
-                payload,
-            });
+        let (local_shutdown, local_shutdown_listener) = oneshot::channel::<()>();
+        let (remote_shutdown, remote_shutdown_listener) = oneshot::channel::<()>();
+        let shutdown_handle = BridgeShutdownHandle {
+            local_shutdown,
+            remote_shutdown,
+        };
 
-            if let Some(f) = forward_publication {
-                debug!("Save message to store");
-                self.inner.push(f).map_err(BridgeError::Store)?;
-            } else {
-                warn!("No topic matched");
+        let local_pump = self.local_pump.run(local_shutdown_listener);
+        let remote_pump = self.remote_pump.run(remote_shutdown_listener);
+        pin_mut!(local_pump);
+        pin_mut!(remote_pump);
+
+        // TODO REVIEW: Do we want to shut down?
+        //           We can either recreate the pump or shut everything down and start over.
+        //
+        //           If there is a client error then this can potentially get reset without the pump shutting down
+        //           Alternatively, if this client error shuts down the pump, we will need to recreate it.
+        debug!(
+            "Starting pumps for {} bridge...",
+            self.connection_settings.name()
+        );
+        match select(local_pump, remote_pump).await {
+            Either::Left(_) => {
+                shutdown_handle.shutdown().await?;
+            }
+            Either::Right(_) => {
+                shutdown_handle.shutdown().await?;
             }
         }
```

Comment:
For context, pump.run() cannot error. That means a shutdown signifies that shutdown has been requested. If we decide to handle client errors by recreating the pump, then we will need to alter this shutdown logic to fit this new behavior.

Reply:
Actually I am wrong, pump.run() I think should error in two cases:
- The client needs to be recreated
- We fail a borrow of the shared state (might not be relevant if we switch to locking)

I think if a pump stops, the cleanest approach is to recreate it.